File: LeetCode/380InsertDeleteGetRandomO(1)/InsertDeleteGetRandomO(1).py
class RandomizedSet:

    def __init__(self):
        self.nums_map = {}
        self.nums_array = []        

    # Time:O(1), Space:O(n)
    def insert(self, val: int) -> bool:
        if val not in self.nums_map:
            self.nums_map[val] = len(self.nums_array)
            self.nums_array.append(val)
            return True
        return False

    # Time:O(1), Space:O(n)
    def remove(self, val: int) -> bool:
        if val in self.nums_map:
            self.nums_array[self.nums_map[val]] = self.nums_array[len(self.nums_array) - 1]
            self.nums_map[self.nums_array[len(self.nums_array) - 1]] = self.nums_map[val]
            del self.nums_map[val]
            self.nums_array.pop()
            return True
        return False        

    # Time:O(1), Space:O(1)
    def getRandom(self) -> int:
        return random.choice(self.nums_array)

File: LeetCode/380InsertDeleteGetRandomO(1)/InsertDeleteGetRandomO(1).cpp
class RandomizedSet {

public:

    std::vector<int> arr;
    std::map<int, int> m;

    RandomizedSet() {
    }
    
    // Time:O(1), Space:O(n)
    bool insert(int val) {
        if (m.find(val) == m.end()) {
            m[val] = arr.size();
            arr.push_back(val);
            return true;
        }
        return false;
    }
    
    // Time:O(1), Space:O(n)
    bool remove(int val) {
        if (m.find(val) != m.end()) {
            arr[m[val]] = arr[arr.size() - 1];
            m[arr[arr.size() - 1]] = m[val];
            m.erase(val);
            arr.pop_back();
            return true;
        }
        return false;
    }
    
    // Time:O(1), Space:O(1)
    int getRandom() {
        return arr[rand() % arr.size()];
    }
};
